Wonderful Sage Goldfish

medium

# Two instances of precision loss in AbstractRewarder leading to irrecoverable funds

## Summary
Precision loss is happening in two function of AbstractRewarder, one is earned() and another is notifyRewardAmount()

## Vulnerability Detail
Lets walk through 2 scenarios to understand - 
1. 
Approx 50,400 block = 7days on mainnet (1 block per 12 seconds).
Lets say 100,000 reward tokens are queued,assume nothing in queue.
```solidity
rewardRate = reward / durationInBlock; ///@audit < rewardRate set in notifyReward
```
so notifyRewardAmount will calculate rewardRate  = 100,000/ 50,400 =1.98412698413 = 1.
So, now totalRewards users will be able to claim = 1 \* 50,400 = 50,400. So the remaining 49,600 tokens will be stuck in abstractRewarder as there is no way to recover them. Here loss is limited to < durationInBlock. Which is not significant in terms of 1e18, but it can add up in a year or two.  

2. 
earned() function calls rewardPerToken
```solidity
function earned(address account) public view returns (uint256) {
return (balanceOf(account) * (rewardPerToken() - userRewardPerTokenPaid[account]) / 1e18) + rewards[account];
}
```
   But in rewardPerToken()
```solidity
uint256 total = totalSupply();
return rewardPerTokenStored + ((lastBlockRewardApplicable() - lastUpdateBlock) * rewardRate * 1e18 / total);///@audit < division here by total
```
division by totalSupply is happening.So effectively in earned() division is happening first and then balance(account) is multiplied **after** the bracket calculations, leading to precision loss. I am attaching a working poc for this scenario below where 100 ether reward is queued & durationInBlock = 100 so that RewardRate = 1 ether so that  first scenario's precision loss does not come into play but you will see funds still getting stuck.
```solidity
    function test_precision_error() public {
        _asset.mint(address(this), 5789 ether);
        _asset.approve(address(_lmpVault), 5789 ether);
        _lmpVault.deposit(5789 ether,address(this));
        _accessController.grantRole(Roles.DV_REWARD_MANAGER_ROLE, address(this));
        _lmpVault.rewarder().addToWhitelist(address(this));
        _toke.mint(address(this), 100 ether);
        _toke.approve(address(_lmpVault.rewarder()), 100 ether);
        _lmpVault.rewarder().queueNewRewards(100 ether);
        vm.roll(block.number + 101);
        _lmpVault.rewarder().getReward();
        emit log_named_decimal_uint("rewards stuck",_toke.balanceOf(address(_lmpVault.rewarder())),18);
        assertGt(_toke.balanceOf(address(_lmpVault.rewarder())),0);
    }
```
add it to LMPVault-Withdraw.t.sol file in LMPVaultMintingTests contract and run using 
```solidity
forge test --match-path ./test/vault/LMPVault-Withdraw.t.sol --match-test test_precision_error -vv
```

Logs:
  rewards stuck: 0.000000000000003747
## Impact

Precision loss in notifyRewardAmount is limited to < blockInDuration. Trivial but can add up.
Precision loss in earned() is erratic but is directly proportional to balanceOf the claiming account. 

## Code Snippet
[earned](https://github.com/sherlock-audit/2023-06-tokemak/blob/5d8e902ce33981a6506b1b5fb979a084602c6c9a/v2-core-audit-2023-07-14/src/rewarders/AbstractRewarder.sol#L204-L206C6)

[rewardPerToken](https://github.com/sherlock-audit/2023-06-tokemak/blob/5d8e902ce33981a6506b1b5fb979a084602c6c9a/v2-core-audit-2023-07-14/src/rewarders/AbstractRewarder.sol#L174-L181)

[NotifyRewardAmount](https://github.com/sherlock-audit/2023-06-tokemak/blob/5d8e902ce33981a6506b1b5fb979a084602c6c9a/v2-core-audit-2023-07-14/src/rewarders/AbstractRewarder.sol#L274-L291)

## Tool used

Manual Review

## Recommendation

Only safety mechanism i can think of without too many changes would be to add a recover rewards function.
